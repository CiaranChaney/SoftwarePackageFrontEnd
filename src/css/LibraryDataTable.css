.upload-container {
    display: flex;
    align-items: center;
    margin-bottom: 15px;
}

.file-input {
    margin-right: 10px;
    padding: 15px;
    border: 1px solid #555;
    border-radius: 5px;
    background-color: #2e3033;
    color: #fff;
}

.upload-button {
    padding: 15px 30px;
    border: none;
    border-radius: 5px;
    background-color: #2e3033;
    color: #fff;
    cursor: pointer;
}

.upload-button:hover {
    background-color: #4b5057;
}
