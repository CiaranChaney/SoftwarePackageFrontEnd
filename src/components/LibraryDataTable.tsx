import React, { useEffect, useState } from "react";
import { DataGrid } from "@mui/x-data-grid";
import {TextField} from "@mui/material";
import axios from "axios";
import moment from "moment";
import { jwtDecode } from "jwt-decode";
import LibraryUpload from "./LibraryUpload";
import ReactGA from "react-ga4";
import { toast, ToastContainer } from "react-toastify";
import 'react-toastify/dist/ReactToastify.css';
import {ThreeCircles} from "react-loader-spinner";
import 'bootstrap/dist/css/bootstrap.css';

ReactGA.initialize('G-QZSYBWE1M5');

ReactGA.send({ hitType: 'pageview', page: "/library"});

declare var localStorage: any;

const token = localStorage.getItem("token");
const decodedToken = token && jwtDecode(token);
const isAdmin =
  decodedToken && decodedToken.roles && decodedToken.roles.includes("ADMIN");

interface Library {
    libraryId: number;
    libraryName: string;
    version: string;
    createdAt: string;
    hash: {
        hashValue: string;
    };
}

const LibraryDataTable = () => {
  const [searchQuery, setSearchQuery] = useState("");
  const [hashSearchQuery, setHashSearchQuery] = useState("");
  const [listLibrary, setListLibrary] = useState<Library[]>([]);
  const [currentPage, setCurrentPage] = useState(0);
  const [pageSize, setPageSize] = useState(50);
  const [totalRows, setTotalRows] = useState(0);
  const [loading, setLoading] = useState(true);

  useEffect(() => {
    const toastMessage = localStorage.getItem("toastMessage");
    if (toastMessage) {
      toast.success(toastMessage, {
        onClose: () => {
          localStorage.removeItem("toastMessage");
        }
      });
    }

    fetchLibraryData();
  }, [currentPage, pageSize]);

  const fetchLibraryData = () => {
    axios
        .get("https://ciaranchaney.com:443/paginatedLibraries", {
          params: {
            pageNo: currentPage,
            pageSize: pageSize
          },
        })
        .then((response) => {
          if (Array.isArray(response.data.content)) {
            setListLibrary(response.data.content);
            setTotalRows(response.data.totalElements);
            setLoading(false)
          } else {
            console.error("Response data is not an array");
          }
        })
        .catch((error) => {
          console.error("There was an error!", error);
        });
  };

  const handlePaginationModelChange = (newModel: { page: React.SetStateAction<number>; pageSize: React.SetStateAction<number>; }) => {
    setCurrentPage(newModel.page);
    setPageSize(newModel.pageSize);
  };

  const handleNameSearch = () => {
    axios
        .get("https://ciaranchaney.com:443/searchLibrary", {
          params: { query: searchQuery },
        })
        .then((response) => {
          if (Array.isArray(response.data)) {
            // @ts-ignore
            setListLibrary(response.data);
          } else {
            console.error("Invalid response data format");
          }
        })
        .catch((error) => {
          console.error("There was an error!", error);
        });
  };

  const handleHashSearch = () => {
    axios
        .get("https://ciaranchaney.com:443/searchByHash", {
          params: { query: hashSearchQuery },
        })
        .then((response) => {
          if (response.data) {
            // @ts-ignore
            setListLibrary([response.data]);
          } else {
            console.error("No library found with that hash");
            setListLibrary([]);
          }
        })
        .catch((error) => {
          console.error("There was an error!", error);
        });
  };


  const columns = [
    { field: "libraryId", headerName: "Library ID", width: 100 },
    { field: "libraryName", headerName: "Library Name", width: 400 },
    { field: "version", headerName: "Version", width: 120 },
    { field: "createdAt", headerName: "Created At", width: 150 },
    {
      field: "hash",
      headerName: "Hash",
      width: 550,
    },
    {
      field: "edit",
      headerName: "",
      width: 100,
      renderCell: (params: { row: { libraryId: any; }; }) => (
        <div>
          {isAdmin && (
            <a className={"btn btn-warning"}
              href={`/editlibrary/${params.row.libraryId}`}
              style={{ color: "#fff" }}
            >
              Edit
            </a>
          )}
        </div>
      ),
      sortable: false,
      filterable: false,
      disableColumnMenu: true,
    },
    {
      field: "delete",
      headerName: "",
      width: 120,
      renderCell: (params: { row: { libraryId: any; }; }) => (
        <div>
          {isAdmin && (
            <a className={"btn btn-danger"}
              href={`/deletelibrary/${params.row.libraryId}`}
              style={{ color: "#fff" }}
            >

              Delete
            </a>
          )}
        </div>
      ),
      sortable: false,
      filterable: false,
      disableColumnMenu: true,
    },
  ];

  const rows = listLibrary.map((library) => ({
    id: library.libraryId,
    libraryId: library.libraryId,
    libraryName: library.libraryName,
    version: library.version,
    createdAt: moment(library.createdAt).format("DD-MM-YYYY"),
    hash: library.hash ? library.hash.hashValue : "",
  }));


  return (
      <div>
        <div className={"col"}>
          <ToastContainer/>
        </div>

        <div className={"row"}>
          <div className={"container-fluid"}>
            <div className={"row"}>
              <div className={"col-lg-4 col-md-6 mb-3"}>
                <TextField
                    label="Search by Library Name"
                    value={searchQuery}
                    onChange={(event) => setSearchQuery(event.target.value)}
                    variant="outlined"
                    fullWidth
                    margin="normal"
                />
                <a className={"btn btn-dark btn-lg mt-2"}
                   onClick={handleNameSearch}
                   role="button">
                  Search by Name
                </a>
              </div>

              <div className={"col-lg-4 col-md-6 mb-3"}>
                <LibraryUpload
                    onFileUpload={(hashValue: React.SetStateAction<string>) => setHashSearchQuery(hashValue)}
                />
              </div>

              <div className={"col-lg-4 col-md-12 mb-3"}>
                <TextField
                    label="Search by Hash"
                    value={hashSearchQuery}
                    onChange={(event) => setHashSearchQuery(event.target.value)}
                    variant="outlined"
                    fullWidth
                    margin="normal"
                />
                <a className={"btn btn-dark btn-lg mt-2"}
                   onClick={handleHashSearch}
                   role="button">
                  Search by Hash
                </a>
              </div>
            </div>

            <div className={"row"}>
              <div className={"col"}>
                <div className={"mt-3 mb-3"}>
                  {loading ? (
                      <ThreeCircles
                          visible={true}
                          height="100"
                          width="100"
                          color="#ffffffff"
                          ariaLabel="three-circles-loading"
                          wrapperStyle={{}}
                          wrapperClass=""
                      />
                  ) : (
                      <div className={"col-10 mx-auto"}>
                        <DataGrid
                            rows={rows}
                            columns={columns}
                            pagination={true}
                            rowCount={totalRows}
                            pageSizeOptions={[25, 50, 100]}
                            initialState={{pagination: {paginationModel: {page: 0, pageSize: 100}}}}
                            paginationMode="server"
                            paginationModel={{page: currentPage, pageSize}}
                            onPaginationModelChange={handlePaginationModelChange}
                        />
                      </div>
                  )}
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>


  );
};

export default LibraryDataTable;
