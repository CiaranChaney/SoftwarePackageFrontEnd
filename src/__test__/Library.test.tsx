import { render, screen, fireEvent } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import LibraryPage from '../pages/Library';
import '@testing-library/jest-dom';
import LibraryDataTable from "../components/LibraryDataTable";

beforeEach(() => {
    jest.doMock("../components/NavBar.tsx", () => {
        return () => (
            <div>
                <nav>
                    <ul>
                        <li>NavBar</li>
                    </ul>
                </nav>
            </div>
        );
    });
    jest.doMock("../components/LibraryDataTable.tsx", () => () => <div>LibraryDataTable</div>);
    jest.doMock("../components/Footer.tsx", () => () => <div>Footer</div>);
});


describe('LibraryPage', () => {
    it('renders correctly', () => {
        const { asFragment } = render(
            <BrowserRouter>
                <LibraryPage />
            </BrowserRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('includes necessary subcomponents', () => {
        const { getByText } = render(
            <BrowserRouter>
                <LibraryPage />
            </BrowserRouter>
        );
        expect(document.querySelector('link[rel="icon"]')).toBeInTheDocument();
        expect(getByText('Library Hash Repository')).toBeInTheDocument();
        expect(getByText('Login')).toBeInTheDocument();
        expect(getByText('Register')).toBeInTheDocument();
        expect(getByText('Contact Us')).toBeInTheDocument();
        expect(getByText('Terms of Service')).toBeInTheDocument();
        expect(getByText('Privacy Policy')).toBeInTheDocument();
    });


    it('renders LibraryDataTable and checks for elements', () => {
        render(<LibraryDataTable />);

        const libraryNameField = screen.getByLabelText('Search by Library Name');
        expect(libraryNameField).toBeInTheDocument();

        const hashField = screen.getByLabelText('Search by Hash');
        expect(hashField).toBeInTheDocument();

        const searchByNameButton = screen.getByRole('button', { name: 'Search by Name' });
        expect(searchByNameButton).toBeInTheDocument();

        const searchByHashButton = screen.getByRole('button', { name: 'Search by Hash' });
        expect(searchByHashButton).toBeInTheDocument();

        fireEvent.change(libraryNameField, { target: { value: 'test library' } });
        fireEvent.click(searchByNameButton);

        fireEvent.change(hashField, { target: { value: 'test hash' } });
        fireEvent.click(searchByHashButton);


    });
});
