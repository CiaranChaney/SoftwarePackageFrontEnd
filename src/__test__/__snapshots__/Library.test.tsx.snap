// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LibraryPage renders correctly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="col"
    >
      <div
        class="container-fluid"
      >
        <div
          class="row"
        >
          <div
            class="navCol col-lg-6 mb-3"
          >
            <div
              class="row"
            >
              <div
                class="col"
              >
                <a
                  class="link-light"
                  href="/"
                >
                  <h4>
                    Library Hash Repository
                  </h4>
                </a>
              </div>
            </div>
          </div>
          <div
            class="navCol col-lg-6 d-flex justify-content-end mb-3"
          >
            <div
              class="row"
            >
              <div
                class="col-9 col-lg-4 col-sm-6 d-flex justify-content-end mb-2 mb-lg-0"
              >
                <a
                  class="btn btn-light btn-lg p-2"
                  href="/library"
                  role="button"
                >
                  Libraries
                </a>
              </div>
              <div
                class="col-3 col-lg-3 col-sm-6 d-flex justify-content-end mb-2 mb-lg-0"
              >
                <a
                  class="btn btn-light btn-lg p-2"
                  href="/login"
                  role="button"
                >
                  Login
                </a>
              </div>
              <div
                class="col-4 col-lg-4 col-sm-6 d-flex justify-content-end mb-2 mb-lg-0"
              >
                <a
                  class="btn btn-light btn-lg p-2"
                  href="/register"
                  role="button"
                >
                  Register
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="col"
      >
        <div
          class="Toastify"
        />
      </div>
      <div
        class="row"
      >
        <div
          class="container-fluid"
        >
          <div
            class="row"
          >
            <div
              class="col-lg-4 col-md-6 mb-3"
            >
              <div
                class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for=":r0:"
                  id=":r0:-label"
                >
                  Search by Library Name
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r0:"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Search by Library Name
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <a
                class="btn btn-dark btn-lg mt-2"
                role="button"
              >
                Search by Name
              </a>
            </div>
            <div
              class="col-lg-4 col-md-6 mb-3"
            >
              <div
                class="col"
              >
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="formFile"
                  />
                  <input
                    class="form-control"
                    id="formFile"
                    type="file"
                  />
                </div>
                <div
                  class="col"
                >
                  <button
                    class="btn btn-dark btn-lg"
                    type="button"
                  >
                    Upload File
                  </button>
                </div>
              </div>
            </div>
            <div
              class="col-lg-4 col-md-12 mb-3"
            >
              <div
                class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
              >
                <label
                  class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                  data-shrink="false"
                  for=":r1:"
                  id=":r1:-label"
                >
                  Search by Hash
                </label>
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r1:"
                    type="text"
                    value=""
                  />
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-yjsfm1"
                    >
                      <span>
                        Search by Hash
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
              <a
                class="btn btn-dark btn-lg mt-2"
                role="button"
              >
                Search by Hash
              </a>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="col"
            >
              <div
                class="mt-3 mb-3"
              >
                <div
                  aria-busy="true"
                  aria-label="three-circles-loading"
                  class="sc-beySPh gZmPOF"
                  data-testid="three-circles-wrapper"
                  role="progressbar"
                >
                  <svg
                    data-testid="three-circles-svg"
                    enable-background="new 0 0 100 100"
                    height="100"
                    version="1.1"
                    viewBox="0 0 100 100"
                    width="100"
                    xml:space="preserve"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M31.6,3.5C5.9,13.6-6.6,42.7,3.5,68.4c10.1,25.7,39.2,38.3,64.9,28.1l-3.1-7.9c-21.3,8.4-45.4-2-53.8-23.3 c-8.4-21.3,2-45.4,23.3-53.8L31.6,3.5z"
                      fill="#ffffffff"
                    >
                      <animatetransform
                        attributeName="transform"
                        attributeType="XML"
                        dur="2s"
                        from="0 50 50"
                        repeatCount="indefinite"
                        to="360 50 50"
                        type="rotate"
                      />
                    </path>
                    <path
                      d="M42.3,39.6c5.7-4.3,13.9-3.1,18.1,2.7c4.3,5.7,3.1,13.9-2.7,18.1l4.1,5.5c8.8-6.5,10.6-19,4.1-27.7 c-6.5-8.8-19-10.6-27.7-4.1L42.3,39.6z"
                      fill="#ffffffff"
                    >
                      <animatetransform
                        attributeName="transform"
                        attributeType="XML"
                        dur="1s"
                        from="0 50 50"
                        repeatCount="indefinite"
                        to="-360 50 50"
                        type="rotate"
                      />
                    </path>
                    <path
                      d="M82,35.7C74.1,18,53.4,10.1,35.7,18S10.1,46.6,18,64.3l7.6-3.4c-6-13.5,0-29.3,13.5-35.3s29.3,0,35.3,13.5 L82,35.7z"
                      fill="#ffffffff"
                    >
                      <animatetransform
                        attributeName="transform"
                        attributeType="XML"
                        dur="2s"
                        from="0 50 50"
                        repeatCount="indefinite"
                        to="360 50 50"
                        type="rotate"
                      />
                    </path>
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="bg-dark text-white text-center py-4"
    >
      <div
        class="container"
      >
        <nav>
          <ul
            class="list-inline mb-0"
          >
            <li
              class="list-inline-item"
            >
              <a
                class="text-white"
                href="/contact"
              >
                Contact Us
              </a>
            </li>
            <li
              class="list-inline-item"
            >
              <a
                class="text-white"
                href="/terms"
              >
                Terms of Service
              </a>
            </li>
            <li
              class="list-inline-item"
            >
              <a
                class="text-white"
                href="/privacy"
              >
                Privacy Policy
              </a>
            </li>
          </ul>
        </nav>
        <p
          class="mb-0"
        >
          © 2024 Library Data Repository. All rights reserved.
        </p>
      </div>
    </footer>
  </div>
</DocumentFragment>
`;
